File: projects/native-apps/ios/App/DrainUploadJobs.swift
import Foundation

@MainActor
struct DrainUploadJobs {
	let api: ReadplaceAPI
	let captor: HTMLCapturing
	let jobs: UploadJobStore
	var now: () -> Date = Date.init

	private static let pdfMediaType = "application/pdf"
	private static let pdfMagic = Data("%PDF-".utf8)

	func run() async {
		jobs.removeOrphanedBytes()
		let due = jobs.loadAll(now: now())
		guard !due.isEmpty else { return }
		guard let page = try? await api.loadQueue() else { return }
		guard let action = page.action(named: "save-content") else {
			for job in due { jobs.remove(job) }
			return
		}
		for job in due where !Task.isCancelled {
			guard await upload(job, through: action) else { return }
		}
	}

	private func upload(_ job: UploadJob, through action: SirenAction) async -> Bool {
		var current = job
		do {
			guard let (ready, contentType) = try await readied(job) else {
				jobs.remove(job)
				return true
			}
			current = ready
			let body = try Data(contentsOf: jobs.bytesURL(for: ready))
			try await api.saveContent(action: action, contentType: contentType, body: body)
			jobs.remove(ready)
		} catch APIError.unauthorized, APIError.noToken {
			return false
		} catch APIError.refused {
			jobs.remove(current)
		} catch APIError.server(let status, _, _) where (400..<500).contains(status) {
			jobs.remove(current)
		} catch {
			reschedule(current)
		}
		return true
	}

	private func readied(_ job: UploadJob) async throws -> (job: UploadJob, contentType: String)? {
		switch job.state {
		case .ready(let contentType):
			return (job, contentType)
		case .capturePending(let detectedMediaType):
			guard let url = URL(string: job.url),
				let form = await content(of: job, at: url, detectedMediaType: detectedMediaType)
			else { return nil }
			return (try await jobs.stageReady(job, form: form), form.contentType)
		}
	}

	private func content(of job: UploadJob, at url: URL, detectedMediaType: String?) async -> MultipartForm? {
		if detectedMediaType == Self.pdfMediaType {
			guard let bytes = await api.fetchExternalContent(url), bytes.starts(with: Self.pdfMagic) else { return nil }
			return saveContentForm(url: url, bytes: bytes, mediaType: Self.pdfMediaType, title: job.title)
		}
		let captured = await captor.capture(url: url)
		guard let html = captured.rawHtml, !html.isEmpty else { return nil }
		let title = (captured.title?.isEmpty == false) ? captured.title : job.title
		return saveContentForm(url: url, bytes: Data(html.utf8), mediaType: "text/html", title: title)
	}

	private func reschedule(_ job: UploadJob) {
		guard let retried = job.retried(now: now()) else { return jobs.remove(job) }
		try? jobs.update(retried)
	}
}

File: projects/native-apps/android/app/src/main/kotlin/com/readplace/android/app/DrainUploadJobs.kt
package com.readplace.android.app

import com.readplace.android.core.ApiError
import com.readplace.android.core.CapturedPage
import com.readplace.android.core.HtmlCapturing
import com.readplace.android.core.MultipartForm
import com.readplace.android.core.ReadplaceApi
import com.readplace.android.core.SirenAction
import com.readplace.android.core.UploadJob
import com.readplace.android.core.UploadJobStore
import com.readplace.android.share.saveContentForm
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.isActive
import java.io.IOException
import java.time.Instant

class DrainUploadJobs(
	private val api: ReadplaceApi,
	private val captor: HtmlCapturing,
	private val jobs: UploadJobStore,
	private val now: () -> Instant,
) {
	private data class Readied(val job: UploadJob, val contentType: String)

	suspend fun run() {
		jobs.removeOrphanedBytes()
		val due = jobs.loadAll(now())
		if (due.isEmpty()) return
		val page = try {
			api.loadQueue()
		} catch (_: ApiError) {
			return
		} catch (_: IOException) {
			return
		}
		val action = page.action(named = "save-content")
		if (action == null) {
			for (job in due) jobs.remove(job)
			return
		}
		for (job in due) {
			if (!currentCoroutineContext().isActive) return
			if (!upload(job, through = action)) return
		}
	}

	private suspend fun upload(job: UploadJob, through: SirenAction): Boolean {
		var current = job
		try {
			val readied = readied(job)
			if (readied == null) {
				jobs.remove(job)
				return true
			}
			current = readied.job
			val body = jobs.bytesFile(readied.job).readBytes()
			api.saveContent(through, readied.contentType, body)
			jobs.remove(readied.job)
		} catch (_: ApiError.Unauthorized) {
			return false
		} catch (_: ApiError.NoToken) {
			return false
		} catch (_: ApiError.Refused) {
			jobs.remove(current)
		} catch (error: ApiError.Server) {
			if (error.status in 400..499) jobs.remove(current) else reschedule(current)
		} catch (_: ApiError) {
			reschedule(current)
		} catch (_: IOException) {
			reschedule(current)
		}
		return true
	}

	private suspend fun readied(job: UploadJob): Readied? =
		when (val state = job.state) {
			is UploadJob.State.Ready -> Readied(job, state.contentType)
			is UploadJob.State.CapturePending -> {
				val form = content(job, detectedMediaType = state.detectedMediaType) ?: return null
				Readied(jobs.stageReady(job, form), form.contentType)
			}
		}

	private suspend fun content(job: UploadJob, detectedMediaType: String?): MultipartForm? {
		if (detectedMediaType == PDF_MEDIA_TYPE) {
			val bytes = api.fetchExternalContent(job.url) ?: return null
			if (!isPdf(bytes)) return null
			return saveContentForm(url = job.url, bytes = bytes, mediaType = PDF_MEDIA_TYPE, title = job.title)
		}
		val captured = captor.capture(job.url)
		if (captured !is CapturedPage.Html || captured.html.isEmpty()) return null
		val title = captured.title?.takeIf { it.isNotEmpty() } ?: job.title
		return saveContentForm(
			url = job.url,
			bytes = captured.html.toByteArray(Charsets.UTF_8),
			mediaType = "text/html",
			title = title,
		)
	}

	private fun reschedule(job: UploadJob) {
		val retried = job.retried(now())
		if (retried == null) {
			jobs.remove(job)
			return
		}
		try {
			jobs.update(retried)
		} catch (_: IOException) {
		}
	}

	private fun isPdf(bytes: ByteArray): Boolean =
		bytes.size >= PDF_MAGIC.size && PDF_MAGIC.indices.all { bytes[it] == PDF_MAGIC[it] }

	private companion object {
		const val PDF_MEDIA_TYPE = "application/pdf"
		val PDF_MAGIC: ByteArray = "%PDF-".toByteArray(Charsets.UTF_8)
	}
}
